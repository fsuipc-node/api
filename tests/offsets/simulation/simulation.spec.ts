import { simulation as offsets } from '@offsets/simulation/simulation';

describe('offset - plane/simulation', () => {
  const offsetsTestCases = [
    { name: 'cloudComplex', value: 0, expectedResult: false },
    { name: 'cloudComplex', value: 1, expectedResult: true },
    { name: 'pauseFlag', value: 0, expectedResult: false },
    { name: 'pauseFlag', value: 1, expectedResult: true },
    { name: 'framerate', value: 354, expectedResult: 92 },
    { name: 'zoomfactor', value: 128, expectedResult: 2 },
    { name: 'fuelBoxFlag', value: 0, expectedResult: false },
    { name: 'fuelBoxFlag', value: 1, expectedResult: true },
    { name: 'comFreqInc', value: 0, expectedResult: 50 },
    { name: 'comFreqInc', value: 1, expectedResult: 25 },
    { name: 'adfFreqInc', value: 0, expectedResult: 1.0 },
    { name: 'adfFreqInc', value: 1, expectedResult: 0.1 },
    { name: 'smokeSystemAvailable', value: 0, expectedResult: false },
    { name: 'smokeSystemAvailable', value: 1, expectedResult: true },
    { name: 'slewMode', value: 0, expectedResult: false },
    { name: 'slewMode', value: 1, expectedResult: true },
    { name: 'replayInAction', value: 0, expectedResult: false },
    { name: 'replayInAction', value: 1, expectedResult: true },
    { name: 'videoRecording', value: 0, expectedResult: false },
    { name: 'videoRecording', value: 1, expectedResult: true },
    { name: 'crashed', value: 0, expectedResult: false },
    { name: 'crashed', value: 1, expectedResult: true },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = offsets[testedOffset.name].convert.replace(new RegExp(/{VAL}/g), testedOffset.value.toString());

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
