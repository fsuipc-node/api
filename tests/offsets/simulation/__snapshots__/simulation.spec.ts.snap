// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/simulation offsets list should have required properties 1`] = `
Object {
  "adfFreqInc": Offset {
    "category": "simulation",
    "convert": "({VAL}) === 0 ? 1.0 : 0.1",
    "description": "ADF freq increment - KHz",
    "length": undefined,
    "mapping": undefined,
    "name": "adfFreqInc",
    "permission": "rw",
    "type": 2,
    "value": 856,
  },
  "availableMemory": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "available FS memory - kb",
    "length": undefined,
    "mapping": undefined,
    "name": "availableMemory",
    "permission": "r",
    "type": 3,
    "value": 588,
  },
  "cloudComplex": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "cloud simple/complex - 0 = simple, 1 = complex",
    "length": undefined,
    "mapping": undefined,
    "name": "cloudComplex",
    "permission": "rw",
    "type": 0,
    "value": 597,
  },
  "cloudCoverDensity": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "cloud cover density - 5-8",
    "length": undefined,
    "mapping": undefined,
    "name": "cloudCoverDensity",
    "permission": "rw",
    "type": 0,
    "value": 596,
  },
  "comFreqInc": Offset {
    "category": "simulation",
    "convert": "({VAL}) === 0 ? 50 : 25",
    "description": "COM freq increment - KHz",
    "length": undefined,
    "mapping": undefined,
    "name": "comFreqInc",
    "permission": "rw",
    "type": 2,
    "value": 856,
  },
  "controlTimer1": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "control timer 1 - FS2002+ - Seconds",
    "length": undefined,
    "mapping": undefined,
    "name": "controlTimer1",
    "permission": "r",
    "type": 8,
    "value": 784,
  },
  "controlTimer2": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "control timer 2 - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "controlTimer2",
    "permission": "r",
    "type": 3,
    "value": 872,
  },
  "crashDetection": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=Ignore Crash, 1=Detect Crash and restart, 2=Detect Crash and show Graph (last is not applicable to FS2002/4)",
    "length": undefined,
    "mapping": undefined,
    "name": "crashDetection",
    "permission": "rw",
    "type": 5,
    "value": 2096,
  },
  "crashDetectionFSX": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=Ignore Crash, 1=Detect Crash",
    "length": undefined,
    "mapping": undefined,
    "name": "crashDetectionFSX",
    "permission": "r",
    "type": 0,
    "value": 2098,
  },
  "crashDetectionFSXAi": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=Ignore Crash, 1=Detect Crash",
    "length": undefined,
    "mapping": undefined,
    "name": "crashDetectionFSXAi",
    "permission": "r",
    "type": 0,
    "value": 2099,
  },
  "crashed": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "crashed",
    "length": undefined,
    "mapping": undefined,
    "name": "crashed",
    "permission": "r",
    "type": 2,
    "value": 2112,
  },
  "elapsedRealTime": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "elasped real time",
    "length": undefined,
    "mapping": undefined,
    "name": "elapsedRealTime",
    "permission": "r",
    "type": 8,
    "value": 1416,
  },
  "elapsedTime": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "simulated flight time - seconds (paused while simulation paused)",
    "length": undefined,
    "mapping": undefined,
    "name": "elapsedTime",
    "permission": "r",
    "type": 8,
    "value": 1192,
  },
  "flightAnalysisMode": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=off 1=landing 2=course tracking 3=manoeuvres",
    "length": undefined,
    "mapping": undefined,
    "name": "flightAnalysisMode",
    "permission": "rw",
    "type": 6,
    "value": 2068,
  },
  "flightModeDisplay": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=off, 1=coods/hdg/speed, 2=fps, 3=all",
    "length": undefined,
    "mapping": undefined,
    "name": "flightModeDisplay",
    "permission": "rw",
    "type": 5,
    "value": 1532,
  },
  "flightPlan": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "current flight plan - FSX only",
    "length": 256,
    "mapping": undefined,
    "name": "flightPlan",
    "permission": "r",
    "type": 11,
    "value": 304,
  },
  "framerate": Offset {
    "category": "simulation",
    "convert": "Math.floor(32768/{VAL})",
    "description": "framerate",
    "length": undefined,
    "mapping": undefined,
    "name": "framerate",
    "permission": "r",
    "type": 5,
    "value": 628,
  },
  "fuelBoxFlag": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "plane in fuel box flag",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelBoxFlag",
    "permission": "r",
    "type": 5,
    "value": 812,
  },
  "internationalEWSetting": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "international E/W setting - 0=east 1=west",
    "length": undefined,
    "mapping": undefined,
    "name": "internationalEWSetting",
    "permission": "rw",
    "type": 2,
    "value": 3484,
  },
  "internationalNSSetting": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "international N/S setting - 2=north 3=south",
    "length": undefined,
    "mapping": undefined,
    "name": "internationalNSSetting",
    "permission": "rw",
    "type": 2,
    "value": 3480,
  },
  "internationalUnit": Offset {
    "category": "simulation",
    "convert": "{VAL} === 0 ? \`us\` : {VAL} === 1 ? \`metric_feet\` : \`metric_meters\`",
    "description": "international units",
    "length": undefined,
    "mapping": undefined,
    "name": "internationalUnit",
    "permission": "rw",
    "type": 2,
    "value": 3096,
  },
  "lastSavedFlightName": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "name of last saved flight",
    "length": 128,
    "mapping": undefined,
    "name": "lastSavedFlightName",
    "permission": "r",
    "type": 11,
    "value": 1024,
  },
  "localTimeString": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "local time string hh:mm:ss",
    "length": 9,
    "mapping": undefined,
    "name": "localTimeString",
    "permission": "r",
    "type": 11,
    "value": 3104,
  },
  "logbookName": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "logbook name - FS2002+",
    "length": 256,
    "mapping": undefined,
    "name": "logbookName",
    "permission": "r",
    "type": 11,
    "value": 300,
  },
  "macroOrLuaParameters": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "macro or LUA parameters",
    "length": undefined,
    "mapping": undefined,
    "name": "macroOrLuaParameters",
    "permission": "rw",
    "type": 6,
    "value": 3436,
  },
  "macroOrLuaString": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "macro or LUA command",
    "length": 40,
    "mapping": undefined,
    "name": "macroOrLuaString",
    "permission": "rw",
    "type": 11,
    "value": 3440,
  },
  "memoryAssignedToFSUIPC": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": undefined,
    "length": undefined,
    "mapping": undefined,
    "name": "memoryAssignedToFSUIPC",
    "permission": "r",
    "type": 2,
    "value": 600,
  },
  "pauseControl": Offset {
    "category": "simulation",
    "convert": "{VAL} ? 1 : 0",
    "description": "pause control (write only)",
    "length": undefined,
    "mapping": undefined,
    "name": "pauseControl",
    "permission": "w",
    "type": 5,
    "value": 610,
  },
  "pauseFlag": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "simulation paused",
    "length": undefined,
    "mapping": undefined,
    "name": "pauseFlag",
    "permission": "r",
    "type": 2,
    "value": 612,
  },
  "realism": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "realism - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "realism",
    "permission": "rw",
    "type": 5,
    "value": 3140,
  },
  "reliability": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "reliability - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "reliability",
    "permission": "rw",
    "type": 5,
    "value": 882,
  },
  "replayInAction": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "replay in action",
    "length": undefined,
    "mapping": undefined,
    "name": "replayInAction",
    "permission": "rw",
    "type": 6,
    "value": 1576,
  },
  "replayTimerCountdown": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "instant replay time left in seconds - controls the playback",
    "length": undefined,
    "mapping": undefined,
    "name": "replayTimerCountdown",
    "permission": "rw",
    "type": 6,
    "value": 1580,
  },
  "scenaryComplexity": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary complexity - 0-5",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryComplexity",
    "permission": "rw",
    "type": 5,
    "value": 2912,
  },
  "scenaryVariableUSRVAR": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary variable USRVAR 0312h",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryVariableUSRVAR",
    "permission": "rw",
    "type": 5,
    "value": 3542,
  },
  "scenaryVariableUSRVAR2": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary variable USRVAR2 0314h",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryVariableUSRVAR2",
    "permission": "rw",
    "type": 5,
    "value": 3544,
  },
  "scenaryVariableUSRVAR3": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary variable USRVAR3 0316h",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryVariableUSRVAR3",
    "permission": "rw",
    "type": 5,
    "value": 3546,
  },
  "scenaryVariableUSRVAR4": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary variable USRVAR4 0318h",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryVariableUSRVAR4",
    "permission": "r",
    "type": 5,
    "value": 3548,
  },
  "scenaryVariableUSRVAR5": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "scenary variable USRVAR5 031Ah",
    "length": undefined,
    "mapping": undefined,
    "name": "scenaryVariableUSRVAR5",
    "permission": "rw",
    "type": 5,
    "value": 3550,
  },
  "simulationRate": Offset {
    "category": "simulation",
    "convert": "+({VAL} / 256).toFixed(2)",
    "description": "international units",
    "length": undefined,
    "mapping": undefined,
    "name": "simulationRate",
    "permission": "rw",
    "type": 2,
    "value": 3098,
  },
  "slewBackwardForwardRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "Slew fwd if-ve, bwd if +ve, 1=very slow ... 127=very fast, -128 fastest forward",
    "length": undefined,
    "mapping": undefined,
    "name": "slewBackwardForwardRate",
    "permission": "rw",
    "type": 1,
    "value": 1515,
  },
  "slewLeftRightRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "Slew left if -ve, right if +ve, 1=very slow ... 127=very fast, -128 fastest leftward",
    "length": undefined,
    "mapping": undefined,
    "name": "slewLeftRightRate",
    "permission": "rw",
    "type": 1,
    "value": 1517,
  },
  "slewMode": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "slew mode activated",
    "length": undefined,
    "mapping": undefined,
    "name": "slewMode",
    "permission": "rw",
    "type": 5,
    "value": 1500,
  },
  "slewModeDisplay": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0=off, 1=coods/hdg/speed, 2=fps, 3=all",
    "length": undefined,
    "mapping": undefined,
    "name": "slewModeDisplay",
    "permission": undefined,
    "type": 5,
    "value": 1524,
  },
  "slewPitchRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "16384=no pitch slew -16384 pitch up, 16384 pitch down, range 0-32767",
    "length": undefined,
    "mapping": undefined,
    "name": "slewPitchRate",
    "permission": "rw",
    "type": 2,
    "value": 1518,
  },
  "slewRollRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "192 gives a 360 roll in about 1 minute - -ve right, +ve left",
    "length": undefined,
    "mapping": undefined,
    "name": "slewRollRate",
    "permission": "rw",
    "type": 2,
    "value": 1508,
  },
  "slewVerticalRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "16384=no alt slew rate - 16383 to 0 increasing slew UP rates, 16385 to 32767 increasing slew DOWN rates",
    "length": undefined,
    "mapping": undefined,
    "name": "slewVerticalRate",
    "permission": "rw",
    "type": 2,
    "value": 1512,
  },
  "slewYawRate": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "Slew mode turns - +ve values are left, -ve are right - 24 takes about 1 minute to complete a 360",
    "length": undefined,
    "mapping": undefined,
    "name": "slewYawRate",
    "permission": "rw",
    "type": 2,
    "value": 1510,
  },
  "smokeSystemAvailable": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "smoke system available - <= FS2000",
    "length": undefined,
    "mapping": undefined,
    "name": "smokeSystemAvailable",
    "permission": "r",
    "type": 5,
    "value": 1492,
  },
  "soundActive": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "soundActive",
    "length": undefined,
    "mapping": undefined,
    "name": "soundActive",
    "permission": "r",
    "type": 2,
    "value": 2852,
  },
  "soundControl": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "soundControl",
    "length": undefined,
    "mapping": undefined,
    "name": "soundControl",
    "permission": "rw",
    "type": 2,
    "value": 2848,
  },
  "startupFlight": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "startup path",
    "length": 256,
    "mapping": undefined,
    "name": "startupFlight",
    "permission": "r",
    "type": 11,
    "value": 36,
  },
  "textureQuality": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "texture quality - 0 to 3",
    "length": undefined,
    "mapping": undefined,
    "name": "textureQuality",
    "permission": "rw",
    "type": 5,
    "value": 3218,
  },
  "thermalVisualisation": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "0 = none, 1 = natural, 2 = schematic",
    "length": undefined,
    "mapping": undefined,
    "name": "thermalVisualisation",
    "permission": "r",
    "type": 0,
    "value": 598,
  },
  "trafficDensityAirline": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "Airline traffic density percent",
    "length": undefined,
    "mapping": undefined,
    "name": "trafficDensityAirline",
    "permission": "rw",
    "type": 0,
    "value": 592,
  },
  "trafficDensityGA": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "General Aviation traffic density percent",
    "length": undefined,
    "mapping": undefined,
    "name": "trafficDensityGA",
    "permission": "rw",
    "type": 0,
    "value": 593,
  },
  "trafficDensityShips": Offset {
    "category": "simulation",
    "convert": undefined,
    "description": "Ships and Ferries traffic density percent",
    "length": undefined,
    "mapping": undefined,
    "name": "trafficDensityShips",
    "permission": "rw",
    "type": 0,
    "value": 594,
  },
  "videoRecording": Offset {
    "category": "simulation",
    "convert": "!!{VAL}",
    "description": "video recording flag - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "videoRecording",
    "permission": "r",
    "type": 0,
    "value": 1888,
  },
  "zoomfactor": Offset {
    "category": "simulation",
    "convert": "Math.floor({VAL}/64)",
    "description": "zoom factor - FS2002+",
    "length": undefined,
    "mapping": undefined,
    "name": "zoomfactor",
    "permission": "r",
    "type": 5,
    "value": 690,
  },
}
`;
